vowpalwabbit/parser.h

```diff
@@ -35,6 +37,46 @@ struct example_initializer
   example* operator()(example* ex);
 };
 
+struct IO_State {
```

Comment:
IO_State isn't encapsulating an entire concept here. Ideally it wrap all the functionality of a queue behind a thread-safe interface, along with any state information you need.

You should add
- A mutex member variable
- `push()` and `pop()` member functions

Reply:
Could the use of `ptr_queue` from `queue.h` be used in `io_state` instead of having new queue/mutex/cv implemented, and mirror the `ready_parsed_examples` queue usage? `ptr_queue` only takes pointers to objects so `io_state::add_to_queue` could create an `io_item` object on heap and `io_state::pop_io_queue` could return a `unique_ptr` to that object, transferring the ownership.

Reply:
Since this PR is intended for a summer project, there are several reasons I chose not to suggest using the queue implementation we have in `queue.h`. Some of these are educational as the queue is an easy way to get your feet wet when learning about threading. I wouldn't recommend using the `ptr_queue` in either case because there workflows where early termination would cause additional complexity to the cleanup.

Both the author and I understand this PR cannot be merged as-is. My hope is that we'll have time to revisit this change once the main focus of the project is completed. Or if we time is too short for that, I hope to continue working with her on the change afterward to get it into a merge-ready state